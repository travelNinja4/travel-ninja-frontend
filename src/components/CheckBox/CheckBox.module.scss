/**
------------------------------------------------------
styles specific to the CheckBox component
------------------------------------------------------
**/

.checkBoxContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;

  input[type='checkbox'] {
    appearance: none;
    width: 20px;
    height: 20px;
    border: 2px solid var(--color-gray-300);
    border-radius: 6px;
    cursor: pointer;
    position: relative;
    transition: all 0.2s ease;

    &:checked {
      background-color: var(--color-indigo);
      border-color: var(--color-indigo);
      box-shadow: 0 0 0 4px var(--color-indigo-alpha-30);
    }

    &:checked::after {
      content: '';
      position: absolute;
      left: 50%;
      top: 40%;
      width: 6px;
      height: 11px;
      border: solid var(--color-white);
      border-width: 0 2px 2px 0;
      transform: translate(-50%, -50%) rotate(45deg);
    }

    &:disabled {
      cursor: not-allowed;
      background-color: var(--color-gray-50);
      border-color: var(--color-gray-200);
      opacity: 0.6;
    }

    &:focus {
      outline: none;
      box-shadow: 0 0 0 3px var(--color-indigo-alpha-40);
    }
  }
}

.errorText {
  color: var(--color-red-500);
  font-size: 0.875rem;
  margin-top: 4px;
}
