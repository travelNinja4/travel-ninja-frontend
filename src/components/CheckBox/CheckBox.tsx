/**
 * Checkbox component is a clickable UI element that allows users to select or deselect an option
 *
 * @example
 * ```tsx
 * import CheckBox from '@src/components/CheckBox'
 *
 * export default function CheckBox() {
 *   return <CheckBox
 *          name="policyCheck"
 *          label={
 *               <>
 *               I agree to the{' '}
 *               <Link href="/terms">
 *                Terms of Service
 *                </Link>{' '}
 *                and <Link href="/privacy">Privacy Policy</Link>
 *                </>
 *          }
 *          checked={true}
 *          onChange={()=>{}}
 *      />;
 * }
 * ```
 */

'use client';

import React, { useEffect, useRef, ReactNode } from 'react';
import Typography, { Tags } from '@/components/Typography';
import styles from './CheckBox.module.scss';

/**
 * Define the props available for the CheckBox component.
 */
interface CheckBoxProps {
  /**
   * The label displayed next to the checkbox.
   * Can be a string, React element, or any valid ReactNode.
   *
   * @example
   * label="Accept terms"
   * @example
   * label={<span>Custom label element</span>}
   */
  label: ReactNode;

  /**
   * The HTML tag to use for rendering the label text.
   * Defaults to `'span'`.
   * Useful for semantic markup (e.g., 'h3', 'p', etc.).
   */
  tag?: Tags;

  /**
   * Name attribute of the checkbox input element.
   * Also used to generate a unique `id` for the checkbox if not provided.
   */
  name?: string;

  /**
   * Value attribute of the checkbox input element.
   * Sent as part of form data when the checkbox is checked.
   */
  value?: string;

  /**
   * Whether the checkbox is required in a form.
   * When `true`, form submission will be blocked until this checkbox is checked.
   */
  required?: boolean;

  /**
   * Whether the checkbox is displayed in an indeterminate state.
   * This is a visual state often used to represent a partially selected group.
   */
  indeterminate?: boolean;
  /**
   * Controlled checked state for the checkbox.
   * Use this for fully controlled components where state is managed externally.
   * @default false
   */
  checked?: boolean;

  /**
   * Initial checked state for the checkbox.
   * Use this for uncontrolled components where the checkbox manages its own state.
   */
  defaultChecked?: boolean;

  /**
   * Whether the checkbox is disabled.
   * When `true`, the checkbox cannot be interacted with.
   */
  disabled?: boolean;

  /**
   * Callback fired when the checked state changes.
   * Will not be called if the checkbox is disabled.
   *
   * @param checked - The new checked state of the checkbox.
   * @returns
   */
  onChange?: (checked: boolean) => void;

  /**
   * Additional CSS class names to apply to the root label element.
   */
  className?: string;

  /**
   * Inline CSS styles to apply to the root label element.
   */
  style?: React.CSSProperties;
}

export default function CheckBox({
  label,
  tag = 'span',
  name,
  value,
  required,
  indeterminate = false,
  checked,
  defaultChecked,
  disabled,
  onChange,
  style,
}: CheckBoxProps) {
  const checkboxId = name || `checkbox-${Math.random()}`;
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (inputRef.current) {
      inputRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate]);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (disabled) {
      return;
    } else if (onChange) {
      onChange(e.target.checked);
    }
  };

  const getLabelContent = () => {
    if (typeof label === 'string' || React.isValidElement(label)) {
      return label;
    }
  };

  return (
    <label
      data-testid="CheckBoxTest"
      htmlFor={checkboxId}
      className={styles.checkBoxContainer}
      style={style}
    >
      <input
        ref={inputRef}
        type="checkbox"
        id={checkboxId}
        name={name}
        value={value}
        required={required}
        checked={checked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        onChange={handleChange}
        className={styles.checkbox}
      />
      {label && <Typography tag={tag}>{getLabelContent()}</Typography>}
    </label>
  );
}
