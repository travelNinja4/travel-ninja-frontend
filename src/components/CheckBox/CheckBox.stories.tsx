import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { useState } from 'react';
/**
import { fn } from 'storybook/test';
*/
import Link from 'next/link';
import CheckBox from './CheckBox';

const meta = {
  title: 'components/CheckBox',
  component: CheckBox,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: {
      control: 'text',
      description: 'Label text or ReactNode for the checkbox',
    },
    tag: {
      control: 'select',
      options: ['span', 'p', 'h1', 'h2', 'h3', 'h4', 'h6'],
      description: 'HTML tag used for the label text',
    },
    checked: {
      control: 'boolean',
      description: 'Whether the checkbox is checked',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the checkbox is disabled',
    },
    indeterminate: {
      control: 'boolean',
      description: 'Whether the checkbox is in indeterminate state',
    },
    onChange: { action: 'changed' },
  },
  args: {
    label: 'Sample label',
    tag: 'span',
    checked: false,
    disabled: false,
    indeterminate: false,
  },
} satisfies Meta<typeof CheckBox>;

export default meta;
type Story = StoryObj<typeof meta>;

// Default interactive story
export const Default: Story = {
  render: (args) => {
    const [checked, setChecked] = useState(false);
    return (
      <CheckBox {...args} checked={checked} onChange={(newChecked) => setChecked(newChecked)} />
    );
  },
};

// With links interactive story
export const WithLinks: Story = {
  render: (args) => {
    const [checked, setChecked] = useState(true);
    return (
      <CheckBox
        {...args}
        label={
          <>
            I agree to the <Link href="/terms">Terms of Service</Link> and{' '}
            <Link href="/privacy">Privacy Policy</Link>
          </>
        }
        checked={checked}
        onChange={(newChecked) => setChecked(newChecked)}
      />
    );
  },
};

// Disabled story stays static
export const Disabled: Story = {
  args: {
    label: 'Disabled checkbox',
    disabled: true,
    checked: false,
  },
};

// Indeterminate interactive story
export const Indeterminate: Story = {
  render: (args) => {
    const [checked, setChecked] = useState(false);
    return (
      <CheckBox
        {...args}
        label="Partially selected"
        indeterminate={!checked && true}
        checked={checked}
        onChange={(newChecked) => setChecked(newChecked)}
      />
    );
  },
};
