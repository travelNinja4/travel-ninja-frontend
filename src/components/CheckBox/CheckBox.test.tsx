import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import CheckBox from './CheckBox';

describe('Home', () => {
  /** Base props for CheckBox **/
  const baseProps = {
    label: 'CheckBox Test',
    name: 'checkboxTest',
    onChange: jest.fn(),
  };

  /**
   * @test
   * Renders the CheckBox component without throwing an error
   * and checks if the CheckBoxTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<CheckBox {...baseProps} />);
    expect(screen.getByTestId('CheckBoxTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the CheckBox component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<CheckBox {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the CheckBox component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<CheckBox {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not call onChange when disabled', () => {
    const onChange = jest.fn();
    render(<CheckBox {...baseProps} disabled onChange={onChange} />);
    screen.getByRole('checkbox').click();
    expect(onChange).not.toHaveBeenCalled();
  });

  it('calls onChange when enabled and clicked', () => {
    const onChange = jest.fn();
    render(<CheckBox {...baseProps} onChange={onChange} />);
    fireEvent.click(screen.getByRole('checkbox'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('sets indeterminate property on checkbox input', () => {
    const { container } = render(<CheckBox {...baseProps} indeterminate />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input.indeterminate).toBe(true);
  });

  it('can render with defaultChecked true', () => {
    render(<CheckBox {...baseProps} defaultChecked />);
    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('sets indeterminate property on checkbox input', () => {
    const { container } = render(<CheckBox {...baseProps} indeterminate />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input.indeterminate).toBe(true);
  });

  it('renders React element as label', () => {
    render(<CheckBox {...baseProps} label={<span>Element Label</span>} />);
    expect(screen.getByText('Element Label')).toBeInTheDocument();
  });

  it('handles non-string, non-element labels gracefully', () => {
    render(<CheckBox {...baseProps} label={123} />);
    expect(screen.getByTestId('CheckBoxTest')).toBeInTheDocument();
  });

  it('renders without a label prop', () => {
    render(<CheckBox name="noLabel" onChange={jest.fn()} label="no label" />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
  });

  it('renders with a custom tag prop', () => {
    render(<CheckBox {...baseProps} tag="h3" />);
    expect(screen.getByText('CheckBox Test').tagName).toBe('H3');
  });

  it('does not set indeterminate when prop is false', () => {
    const { container } = render(<CheckBox {...baseProps} indeterminate={false} />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input.indeterminate).toBe(false);
  });

  it('renders without a label when label is null', () => {
    render(<CheckBox name="noLabel" onChange={jest.fn()} label={null} />);
    expect(screen.queryByText('CheckBox Test')).not.toBeInTheDocument();
  });
});
