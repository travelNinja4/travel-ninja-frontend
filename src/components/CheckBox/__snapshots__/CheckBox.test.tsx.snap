// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <label
    data-testid="CheckBoxTest"
    for="checkboxTest"
  >
    <input
      class="checkbox"
      id="checkboxTest"
      name="checkboxTest"
      type="checkbox"
    />
    <span
      class="typography span align-left"
      data-testid="TypographyTest"
    >
      CheckBox Test
    </span>
  </label>
</DocumentFragment>
`;
